pkg/daemon/podprobe/prober.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2022 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podprobe
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"time"
+
+	appsv1alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
+	criapi "k8s.io/cri-api/pkg/apis"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/probe"
+	execprobe "k8s.io/kubernetes/pkg/probe/exec"
+	"k8s.io/utils/exec"
+)
+
+const (
+	maxProbeRetries  = 3
+	maxMessageLength = 100
+)
+
+// Prober helps to check the probe(exec, http, tcp) of a container.
+type prober struct {
+	exec           execprobe.Prober
+	runtimeService criapi.RuntimeService
+}
+
+// NewProber creates a Prober, it takes a command runner and
+// several container info managers.
+func newProber(runtimeService criapi.RuntimeService) *prober {
+	return &prober{
+		exec:           execprobe.New(),
+		runtimeService: runtimeService,
+	}
+}
+
+// probe probes the container.
+func (pb *prober) probe(p *appsv1alpha1.ContainerProbeSpec, container *runtimeapi.Container, containerID string) (appsv1alpha1.ProbeState, string, error) {
+	result, msg, err := pb.runProbeWithRetries(p, container, containerID)
+	if bytes.Count([]byte(msg), nil)-1 > maxMessageLength {
+		msg = msg[:maxMessageLength]
+	}
+	if err != nil || (result != probe.Success && result != probe.Warning) {
+		return appsv1alpha1.ProbeFailed, msg, err
```

Comment:
consider record event for probing result

Reply:
record event to pod?

Reply:
Is event necessary here? Currently kruise daemon prints the log.